hal/halx86/include/ioapic.h

```diff
@@ -74,6 +84,35 @@ typedef struct _IOAPIC_INFO
 } IOAPIC_INFO, *PIOAPIC_INFO;
 
 #define IOAPIC_DEFAULT_BASE   0xFEC00000    /* Default I/O APIC Base Register Address */
+#define IOAPIC_SIZE           0x400
+
+#ifndef _M_AMD64
+FORCEINLINE
+ULONG
+IoApicRead(_In_ PIO_APIC_REGISTERS IoApicRegs,
+           _In_ UCHAR RegisterIdx)
+{
+    ULONG Result;
+
+    *(volatile PUCHAR)((ULONG_PTR)IoApicRegs + IOAPIC_IOREGSEL) = RegisterIdx;
+    Result = *(volatile PULONG)((ULONG_PTR)IoApicRegs + IOAPIC_IOWIN); 
+
+    //DbgPrint("IoApicRead: RegisterIdx %X, Result %X\n", RegisterIdx, Result);
+    return Result;
+}
+
+FORCEINLINE
+VOID
+IoApicWrite(_In_ PIO_APIC_REGISTERS IoApicRegs,
+            _In_ UCHAR RegisterIdx,
+            _In_ ULONG Value)
+{
+    //DbgPrint("IoApicWrite: RegisterIdx %X, Value %X\n", RegisterIdx, Value);
+
+    *(volatile PUCHAR)((ULONG_PTR)IoApicRegs + IOAPIC_IOREGSEL) = RegisterIdx;
+    *(volatile PULONG)((ULONG_PTR)IoApicRegs + IOAPIC_IOWIN) = Value;
```

Comment:
Not using WRITE_REGISTER_ULONG or stuff?

Reply:
It is done to avoid function call

Reply:
But when used within the HAL itself, these are not functions:
https://git.reactos.org/?p=reactos.git;a=blob;f=sdk/include/ddk/ioaccess.h;hb=40f7e863472d2b0a662c06a95316997ab7c62d2b#l31

Reply:
I've checked halpc98 DLL using dependency walker and I see *_REGISTER_UCHAR in the list.

Reply:
Yes of course, macros forms are used within the HAL, but then it also needs to export these as functions for the other kernel modules to use them.

Reply:
Interesting, so the NTOSKRNL exports the WRITE/READ_REGISTER_xxx, but it's the HAL that exports the WRITE/READ_PORT_xxx 😆 .
Anyway, if you use the header file I mentioned above, you should be able to use these things as macros.

Reply:
Yes, with "ioaccess.h"it ok.